### PR TITLE
Outgoing traffic

### PR DESCRIPTION
**Pull request by meowjesty**

### CODE REVIEW

File: mirrord-agent/src/tcp/outgoing.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        trace!("intercept_task -> id {:#?}", connection_id);
+
+        let mut buffer = vec![0; 1500];
+        let (remote_reader, mut remote_writer) = stream.into_split();
+
+        loop {
+            select! {
+                biased;
+
+                readable = remote_reader.readable() => {
```
Comment (aviramha) [inline mirrord-agent/src/tcp/outgoing.rs:78]: I think you can use `AsyncReadExt` to avoid the readable, match flow (as it does it for you)

File: mirrord-layer/src/socket/hooks.rs [hunk 2]
```diff
@@ -24,12 +24,44 @@ pub(crate) unsafe extern "C" fn socket_detour(
...
+        debug!("socket_detour -> bypassed");
+        socket_result
+    } else {
+        let (Ok(result) | Err(result)) =
+            socket(socket_result, domain, type_, protocol).map_err(From::from);
+        result
+    }
+}
+
+#[hook_fn]
```
Comment (aviramha) [inline mirrord-layer/src/socket/hooks.rs:38]: I don't think we need this hook?

Comment (meowjesty) [reply mirrord-layer/src/socket/hooks.rs]: Removed, this was just something that I was investigating.

File: mirrord-agent/src/tcp/outgoing.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                // [layer] -> [agent]
+                request = request_rx.recv() => {
+                    trace!("interceptor_task -> request {:?}", request);
+
+                    match request {
+                        Some(request) => {
+                            match request {
+                                // [user] -> [layer] -> [agent] -> [layer]
+                                // `user` is asking us to connect to some remote host.
+                                TcpOutgoingRequest::Connect(ConnectRequest { remote_address }) => {
```
Comment (aviramha) [inline mirrord-agent/src/tcp/outgoing.rs:80]: I think the Request/Response terminology might be confusing (of the enum, `ConnectRequest` is fine). I'd call it `LayerTcpOutgoing` and `DaemonTcpOutgoing` or smth like that?

File: mirrord-agent/src/tcp/outgoing.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                                    {
+                                        Ok(writer) => writer
+                                            .write_all(&bytes)
+                                            .await
+                                            .map_err(ResponseError::from)
+                                            .map(|()| WriteResponse { connection_id }),
+                                        Err(fail) => Err(fail),
+                                    };
+
+                                    let response = TcpOutgoingResponse::Write(write_response);
```
Comment (aviramha) [inline mirrord-agent/src/tcp/outgoing.rs:126]: why do you write back write_response? if it fails to write socket will be terminated anyway?

Comment (meowjesty) [reply mirrord-agent/src/tcp/outgoing.rs:126]: We can get away without it, as you pointed out to me, but as I had already implemented this message flow, why remove it?

File: mirrord-agent/src/tcp/outgoing.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord-agent/src/tcp/outgoing.rs:80]: Sure, I wanted to unify terminology for layer -> agent, and agent -> layer messages. As I was using this request/response for the features I've been implementing, so I went ahead and used the same here.

If we're going to change it to be layer/daemon instead of request/response, it's fine for me, but we should change it everywhere as well.

Comment (meowjesty) [reply mirrord-agent/src/tcp/outgoing.rs]: Changed it to be Layer/Daemon, instead of Request/Response.

File: mirrord-layer/src/tcp/outgoing.rs [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                            warn!("interceptor_task -> exiting due to local stream closed!");
+                            break;
+                        },
+                        Ok(read_amount) => {
+                            // Sends the message that the user wrote to our interceptor socket to
+                            // be handled on the `agent`, where it'll be forwarded to the remote.
+                            let write = Write { connection_id, bytes: buffer[..read_amount].to_vec() };
+                            let outgoing_data = TcpOutgoing::Write(write);
+
+                            if let Err(fail) = send_hook_message(HookMessage::TcpOutgoing(outgoing_data)).await {
```
Comment (aviramha) [inline mirrord-layer/src/tcp/outgoing.rs:125]: Why is this hook message? We're running from the `mirrord-layer` `Runtime` as a task here (AFAIK) - no need to use "bad" synch mechanisms - just pass a sender to this task from the main thread and bridge it.

Comment (aviramha) [reply mirrord-layer/src/tcp/outgoing.rs:125]: (Hook messages should be used from hook context, not tokio..)

File: mirrord-layer/src/tcp/outgoing.rs [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        trace!("handle_hook_message -> message {:?}", message);
+
+        match message {
+            TcpOutgoing::Connect(Connect {
+                remote_address,
+                channel_tx,
+            }) => {
+                trace!("Connect -> remote_address {:#?}", remote_address);
+
+                self.connect_queue.push_back(channel_tx);
```
Comment (aviramha) [inline mirrord-layer/src/tcp/outgoing.rs:182]: Starting to get afraid by the way we do this queues instead of some sort of HashMap (there's an odd case for a race condition probably right now). Nothing that should be done right now, just a thought on the first thing we should refactor - some generic API for passing Request/Response between layer/agent that is ID based for avoiding evil

File: mirrord-layer/src/lib.rs [hunk 7]
```diff
@@ -386,33 +364,56 @@ fn enable_hooks(enabled_file_ops: bool, enabled_remote_dns: bool) {
     interceptor.end_transaction();
 }
 
+// TODO: When this is annotated with `hook_guard_fn`, then the outgoing sockets never call it (we
+// just bypass). Everything works, so, should we intervene?
+//
```
Comment (aviramha) [inline mirrord-layer/src/lib.rs:369]: As far as I follow (AFAIF?) close will close our "proxy" socket, which will then `layer` will see it in the `interceptor_task` and then will notify the remote agent. sounds good. maybe write this flow?

File: mirrord-layer/src/lib.rs [hunk 8]
```diff
@@ -222,6 +180,20 @@ where
...
+                .await
+                .unwrap(),
+            HookMessage::Exit(message) => {
+                trace!("HookMessage::Exit");
+
+                let request = ClientMessage::ExitRequest;
+
+                self.codec.send(request).await.unwrap();
+
+                message.hook_channel_tx.send(Ok(())).unwrap();
```
Comment (aviramha) [inline mirrord-layer/src/lib.rs:195]: break loop for graceful exit here

Comment (aviramha) [reply mirrord-layer/src/lib.rs:195]: ah we need to send a message when we close.. ermm.. we could call exit on our own (bypassing the hook becuase of our bypass) when loop ends..?

File: tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs [hunk 9]
```diff
@@ -0,0 +1,84 @@
+import https from "node:https";
```
Comment (aviramha) [inline tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs:1]: one fear I have for all tests added - how do you know it's actually hooked? like, if layer wouldn't load this test will still pass. not sure if worth investing, but maybe add a comment for now.

**Review by aviramha — CHANGES_REQUESTED** [review]
Overall looks great - awesome job.
Generally I saw some changes that were peripheral to the feature, so I'd suggest to break those into more PRs next time for easier review (might have missed something in those areas because hard to follow if related to the "big change" or not) 
See inline comments.

File: mirrord-layer/src/lib.rs [hunk 10]
```diff
@@ -146,6 +96,7 @@ fn init() {
     };
 
     let enabled_file_ops = ENABLED_FILE_OPS.get_or_init(|| config.enabled_file_ops);
+    let _ = ENABLED_TCP_OUTGOING.get_or_init(|| config.enabled_tcp_outgoing);
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-layer/src/lib.rs:99]: This looks like the better init function for `OnceLock`, the other being [`set`](https://doc.rust-lang.org/nightly/std/sync/struct.OnceLock.html#method.set) which may block.

File: mirrord-agent/src/tcp/outgoing.rs [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+                            let daemon_read = read
+                                .map_err(ResponseError::from)
+                                .map(|bytes| DaemonRead { connection_id, bytes: bytes.to_vec() });
+
+                            let daemon_message = DaemonTcpOutgoing::Read(daemon_read);
+                            daemon_tx.send(daemon_message).await?
+                        }
+                        None => {
+                            trace!("interceptor_task -> close connection {:#?}", connection_id);
+                            let daemon_message = DaemonTcpOutgoing::Close(connection_id);
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-agent/src/tcp/outgoing.rs:166]: Manually keeping in sync now, added a comment to improve this in the future.

File: mirrord-agent/src/sniffer.rs [hunk 12]
```diff
@@ -8,7 +8,7 @@ use std::{
 use futures::StreamExt;
 use mirrord_protocol::{
     tcp::{DaemonTcp, NewTcpConnection, TcpClose, TcpData},
-    ConnectionID, Port,
+    ConnectionId, Port,
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-agent/src/sniffer.rs:11]: Yeah, my bad! I had added a `ConnectionId` type alias as well when working on the feature, as the existing one was using a lower uint range (and I missed it for quite a while ...).

File: mirrord-layer/src/socket/ops.rs [hunk 13]
```diff
@@ -196,8 +239,80 @@ pub(super) fn connect(sockfd: RawFd, remote_address: SocketAddr) -> Result<()> {
...
+            if connect_result == -1 && err_code != libc::EINPROGRESS && err_code != libc::EINTR {
+                error!(
+                    "connect -> Failed call to libc::connect with {:#?} errno is {:#?}",
+                    connect_result,
+                    errno::errno()
+                );
+                return Err(io::Error::from_raw_os_error(connect_result).into());
+            }
+
+            let connected = Connected {
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-layer/src/socket/ops.rs:277]: Added a comment there.

File: mirrord-layer/src/tcp/outgoing.rs (hunk 6 above)
Comment (meowjesty) [reply mirrord-layer/src/tcp/outgoing.rs:184]: `TODO` added.

File: tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs (hunk 9 above)
Comment (meowjesty) [reply tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs:1]: Makes sense, I was wondering about a nice way of testing these, but only came up with the simple ways of doing it.

One thing I've been thinking that we could do, is have some sort of commands that we can send back and forth between layer and agent, to be used internally by us (mostly for tests). Like how you guys added a phase check for the pod status, but more general.

File: mirrord-layer/src/lib.rs (hunk 8 above)
Comment (meowjesty) [reply mirrord-layer/src/lib.rs:195]: Both ways are pretty valid (imo), so:

1. Should I change it now?
2. Keep as-is is fine?
3. Keep as-is, but add a `TODO`?

Comment (aviramha) [reply mirrord-layer/src/lib.rs:195]: I'm okay with 3, just add a comment.

File: mirrord-layer/src/tcp/outgoing.rs (hunk 5 above)
Comment (meowjesty) [reply mirrord-layer/src/tcp/outgoing.rs]: Makes sense, changed it to separate channels.

File: mirrord-agent/src/main.rs [hunk 14]
```diff
@@ -220,17 +235,36 @@ impl ClientConnectionHandler {
...
                         error!("tcp sniffer stopped?");
                         break;
                     }
                 }
+                outgoing_response = self.tcp_outgoing_api.daemon_message() => {
+                    match outgoing_response {
+                        Ok(response) => self.respond(DaemonMessage::TcpOutgoing(response)).await?,
+                        Err(fail) => {
+                            match fail {
+                                AgentError::ReceiverClosed => if self.keep_alive {
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-agent/src/main.rs:254]: It was the easiest way I thought of to keep agent alive when a single request completes. If we don't check for some sort of `keep_alive` thingy, then it was just `break`ing the loop. If I left this case (of not having any requests left) as `continue`, then we would loop forever (encountered both cases while writing the code).

File: mirrord-layer/src/lib.rs (hunk 7 above)
Comment (meowjesty) [reply mirrord-layer/src/lib.rs:383]: Added a flow for closing it on `agent`.

File: tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs (hunk 9 above)
Comment (aviramha) [reply tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs:1]: Don't forget to add todo there so it's clear :)

Comment (meowjesty) [reply tests/node-e2e/outgoing/test_outgoing_traffic_make_request_after_listen.mjs:1]: Added a comment about this in `sanity.rs` before the outgoing traffic tests.

File: mirrord-agent/src/main.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-agent/src/main.rs]: Removed the `exit` hook.

**Review by eyalb181 — APPROVED** [review]

**Review by aviramha — APPROVED** [review]
